src/topology/config/validation.rs

```diff
@@ -17,278 +25,247 @@ enum Node {
+                    | (Node::Source { ty: ty1 }, Node::Transform { in_ty: ty2, .. })
+                    | (Node::Transform { out_ty: ty1, .. }, Node::Transform { in_ty: ty2, .. })
+                    | (Node::Transform { out_ty: ty1, .. }, Node::Sink { ty: ty2, .. }) => {
+                        if ty1 != ty2 && ty2 != DataType::Any {
+                            errors.push(format!(
+                                "Data type mismatch between {} ({:?}) and {} ({:?})",
+                                x, ty1, y, ty2
+                            ));
+                        }
+                    }
+                    (Node::Sink { .. }, _) | (_, Node::Source { .. }) => unreachable!(),
+                }
             }
-            match (nodes[&x].clone(), nodes[&y].clone()) {
-                (Node::Source { ty: ty1 }, Node::Sink { ty: ty2, .. })
-                | (Node::Source { ty: ty1 }, Node::Transform { in_ty: ty2, .. })
-                | (Node::Transform { out_ty: ty1, .. }, Node::Transform { in_ty: ty2, .. })
-                | (Node::Transform { out_ty: ty1, .. }, Node::Sink { ty: ty2, .. }) => {
-                    if ty1 != ty2 {
-                        errors.push(format!(
-                            "Data type mismatch between {} ({:?}) and {} ({:?})",
-                            x, ty1, y, ty2
-                        ));
+        }
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            errors.sort();
+            errors.dedup();
+            Err(errors)
+        }
+    }
+
+    fn edges(&self) -> HashSet<(String, String)> {
+        let mut edges = HashSet::new();
+        for (name, node) in self.nodes.iter() {
+            match node {
+                Node::Transform { inputs, .. } | Node::Sink { inputs, .. } => {
+                    for i in inputs {
+                        edges.insert((i.clone(), name.clone()));
                     }
                 }
-                (Node::Sink { .. }, _) | (_, Node::Source { .. }) => unreachable!(),
+                Node::Source { .. } => {}
             }
         }
+        edges
     }
 
-    if errors.is_empty() {
-        Ok(())
-    } else {
-        errors.sort();
-        errors.dedup();
-        Err(errors)
+    // Modified version of Kahn's topological sort algorithm that ignores the actual sorted output and
+    // only cares if the sort was possible (i.e. whether or not there was a cycle in the input graph).
+    fn contains_cycle(&self) -> bool {
+        let nodes = self
+            .nodes
+            .keys()
+            .map(|k| k.to_string())
+            .collect::<HashSet<_>>();
+        let mut edges = self.edges();
+
+        let mut no_incoming = nodes
+            .into_iter()
+            .filter(|n| !edges.iter().any(|(_t, h)| h == n))
+            .collect::<Vec<String>>();
+        while let Some(node) = no_incoming.pop() {
+            let outgoing = edges
+                .clone()
```

Comment:
I dont think you need to clone this entire set, you should be able to `iter -> filter -> map(Clone::clone) -> collect` I think.

Reply:
Not super worried about performance in this path, but nice catch.